Sources/FoundationEssentials/AttributedString/DiscontiguousAttributedSubstring.swift

```diff
@@ -0,0 +1,320 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if FOUNDATION_FRAMEWORK
+@_spi(Unstable) internal import CollectionsInternal
+#elseif canImport(_RopeModule)
+internal import _RopeModule
+#elseif canImport(_FoundationCollections)
+internal import _FoundationCollections
+#endif
+
+@dynamicMemberLookup
+@available(FoundationPreview 6.2, *)
+public struct DiscontiguousAttributedSubstring: Sendable {
+    /// The guts of the base attributed string.
+    internal var _guts: AttributedString.Guts
+    
+    internal var _indices: RangeSet<BigString.Index>
+    
+    internal var _identity: Int = 0
+    
+    internal init(_ guts: AttributedString.Guts, in indices: RangeSet<BigString.Index>) {
+        self._guts = guts
+        // Forcibly resolve bounds and round them down to nearest scalar boundary.
+        var ranges = Array(indices.ranges)
+        for i in ranges.indices {
+            let slice = _guts.string.unicodeScalars[ranges[i]]
+            ranges[i] = Range(uncheckedBounds: (slice.startIndex, slice.endIndex))
+        }
+        self._indices = RangeSet(ranges)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension DiscontiguousAttributedSubstring {
+    public var base: AttributedString {
+        return AttributedString(_guts)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension DiscontiguousAttributedSubstring : CustomStringConvertible {
+    public var description: String {
+        _guts.description(in: _indices)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension DiscontiguousAttributedSubstring : Equatable {
+    public static func == (lhs: Self, rhs: Self) -> Bool {
+        AttributedString.Guts._characterwiseIsEqual(lhs.runs, to: rhs.runs)
```

Comment:
Is this equivalent to looping over the ranges, and compare their characters one by one, like this?
```
for range in  _indices.ranges {
    AttributedString.Guts.characterwiseIsEqual(lhs.guts, in: range, to: rhs.guts, in: range)
}
```

Asking because in the `hash` implementation below, you're doing it range by range. That looks like a different way to think about this.